adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,712 @@
+            tls_io_instance->on_io_open_complete_context = on_io_open_complete_context;
+
+            tls_io_instance->on_bytes_received = on_bytes_received;
+            tls_io_instance->on_bytes_received_context = on_bytes_received_context;
+
+            tls_io_instance->on_io_error = on_io_error;
+            tls_io_instance->on_io_error_context = on_io_error_context;
+
+            tls_io_instance->tlsio_state = TLSIO_STATE_OPENING;
+
+            if (create_openssl_instance(tls_io_instance) != 0)
+            {
+
+                tls_io_instance->tlsio_state = TLSIO_STATE_ERROR;
+                result = __LINE__;
+            }
+            else
+            {
+                send_handshake_bytes(tls_io_instance);
+                result = 0;
+            }
+        }
+    }
+
+    return result;
+}
+
+
+/* Codes_SRS_TLSIO_SSL_ESP8266_99_013: [ The tlsio_openssl_close succeed.]*/
+int tlsio_openssl_close(CONCRETE_IO_HANDLE tls_io, ON_IO_CLOSE_COMPLETE on_io_close_complete, void* callback_context)
+{
+    int result = 0;
+
+    /* Codes_SRS_TLSIO_SSL_ESP8266_99_011: [ The tlsio_openssl_close NULL parameter.]*/
+    if (tls_io == NULL)
+    {
+        result = __LINE__;
+        LogError("NULL tls_io.");
+    }
+    else
+    {
+        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
+
+        if ((tls_io_instance->tlsio_state == TLSIO_STATE_NOT_OPEN) ||
+            (tls_io_instance->tlsio_state == TLSIO_STATE_CLOSING))
+        {
+            /* Codes_SRS_TLSIO_SSL_ESP8266_99_012: [ The tlsio_openssl_close wrong state.]*/
+            result = __LINE__;
+            LogError("Invalid tlsio_state. Expected state is TLSIO_STATE_NOT_OPEN or TLSIO_STATE_CLOSING.");
+        }
+        else
+        {
+            tls_io_instance->tlsio_state = TLSIO_STATE_CLOSING;
+            tls_io_instance->on_io_close_complete = on_io_close_complete;
+            tls_io_instance->on_io_close_complete_context = callback_context;
+
+            destroy_openssl_instance(tls_io_instance);
+            result = 0;
+        }
+    }
+
+    return result;
+}
+
+int tlsio_openssl_send(CONCRETE_IO_HANDLE tls_io, const void* buffer, size_t size, ON_SEND_COMPLETE on_send_complete, void* callback_context)
+{
+    int result;
+
+    if (tls_io == NULL)
+    {
+        /* Codes_SRS_TLSIO_SSL_ESP8266_99_014: [ The tlsio_openssl_send NULL instance.]*/
+        result = __LINE__;
+        LogError("NULL tls_io.");
+    }
+    else
+    {
+        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
+
+        if (tls_io_instance->tlsio_state != TLSIO_STATE_OPEN)
+        {
+            /* Codes_SRS_TLSIO_SSL_ESP8266_99_015: [ The tlsio_openssl_send wrog state.]*/
+            result = __LINE__;
+            LogError("Invalid tlsio_state. Expected state is TLSIO_STATE_OPEN.");
+        }
+        else
+        {
+            int total_write = 0;
+            int res = 0;
+            int retry = 0;
+
+            while(size > 0 && retry < 500){
```

Comment:
Sounds me that 500 is too much. Please define a constant for the maximum send retries, and consider a smaller value, probably near 10. Consider that if SSL_write returns 0 is because something is wrong.